contracts/MarginCalculator.sol

```diff
@@ -43,19 +36,227 @@ contract MarginCalculator {
+    }
+
+    /**
+     * @notice set new time to expiry for specific product
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _timeToExpiry option time to expiry timestamp
+     */
+    function setProductTimeToExpiry(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _timeToExpiry
+    ) external onlyOwner {
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+        uint256[] storage expiryArray = productTimeToExpiry[productHash];
+
+        require(
+            (expiryArray.length == 0) || (_timeToExpiry > expiryArray[expiryArray.length.sub(1)]),
+            "MarginCalculator: expiry array is not in order"
+        );
+        require(
+            timeToExpiryValue[productHash][_timeToExpiry] != 0,
+            "MarginCalculator: no expiry upper bound value found"
+        );
+
+        expiryArray.push(_timeToExpiry);
+    }
+
+    /**
+     * @notice set option upper bound value for specific time to expiry (1e27)
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _timeToExpiry option time to expiry timestamp
+     * @param _value upper bound value
+     */
+    function setTimeToExpiryValue(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _timeToExpiry,
+        uint256 _value
+    ) external onlyOwner {
+        require(_value > 0, "MarginCalculator: invalid option upper bound value");
+
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+
+        timeToExpiryValue[productHash][_timeToExpiry] = _value;
+    }
+
+    /**
+     * @notice set spot shock value (1e27)
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _shockValue spot shock value
+     */
+    function setSpotShock(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _shockValue
+    ) external onlyOwner {
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+
+        spotShock[productHash] = _shockValue;
+    }
+
+    /**
+     * @notice set oracle deviation (1e27)
+     * @dev can only be called by owner
+     * @param _deviation deviation value
+     */
+    function setOracleDeviation(uint256 _deviation) external onlyOwner {
+        oracleDeviation = _deviation;
+    }
```

Comment:
should oracle deviation be one value for each collateral?

Reply:
It is set as one value for all in the spec.